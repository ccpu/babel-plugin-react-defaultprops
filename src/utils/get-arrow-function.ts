import { types as t } from '@babel/core';

export const getArrowFunction = (
  node: t.CallExpression | t.ArrowFunctionExpression,
) => {
  if (t.isArrowFunctionExpression(node)) {
    return node;
  }

  if (
    t.isMemberExpression(node.callee) &&
    t.isIdentifier(node.callee.object) &&
    t.isIdentifier(node.callee.property)
  ) {
    if (
      node.callee.property.name === 'forwardRef' ||
      node.callee.property.name === 'memo'
    ) {
      const func = node.arguments[0];
      if (t.isArrowFunctionExpression(func)) {
        return func;
      } else if (t.isCallExpression(func)) {
        return getArrowFunction(func);
      }
    }
  }
  return undefined;
};
