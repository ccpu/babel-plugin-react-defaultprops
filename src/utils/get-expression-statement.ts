import { PathInfo } from '../typings';
import { types as t } from '@babel/core';

export const getExpressionStatement = (info: PathInfo) => {
  return t.expressionStatement(
    t.assignmentExpression(
      '=',
      t.memberExpression(
        t.identifier(info.componentName),
        t.identifier('__defaultProps'),
      ),
      t.objectExpression(
        info.props && info.props.length
          ? info.props.map((prop) => {
              return t.objectProperty(prop.left as t.Identifier, prop.right);
            })
          : [],
      ),
    ),
  );
};
