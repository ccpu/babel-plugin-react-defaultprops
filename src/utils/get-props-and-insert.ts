import { getProps, getExpressionStatement, getPropsFormBody } from '.';
import { types as t, NodePath } from '@babel/core';

export const getPropsAndInsert = (
  path: NodePath<t.FunctionDeclaration> | NodePath<t.VariableDeclaration>,
  func: t.FunctionDeclaration,
  componentName: string,
) => {
  const firstParam = func.params.length && func.params[0];

  if (
    firstParam &&
    (t.isObjectExpression(firstParam) || t.isObjectPattern(firstParam))
  ) {
    const exp = getExpressionStatement({
      componentName,
      // node,
      path,
      props: getProps(func),
    });
    path.insertAfter(exp);
  } else if (func.body.body) {
    const props = getPropsFormBody(func);

    const exp = getExpressionStatement({
      componentName,
      // node,
      path,
      props: props || [],
    });

    func.body.body = func.body.body.reduce((arr, node) => {
      if (t.isReturnStatement(node)) {
        arr.push(exp);
      }
      arr.push(node);
      return arr;
    }, []);
  }
};
