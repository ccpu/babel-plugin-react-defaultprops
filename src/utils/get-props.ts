import { types as t } from '@babel/core';

export const getProps = (node: t.FunctionDeclaration) => {
  const properties =
    node.params && t.isObjectPattern(node.params[0])
      ? node.params[0].properties
      : undefined;

  const props = properties.reduce((arr, n) => {
    if (!t.isObjectProperty(n) || !t.isAssignmentPattern(n.value)) {
      return arr;
    }
    arr.push(n.value);
    return arr;
  }, []);

  return props;
};
