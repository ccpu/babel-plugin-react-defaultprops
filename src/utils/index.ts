export * from './is-component';
export * from './get-props';
