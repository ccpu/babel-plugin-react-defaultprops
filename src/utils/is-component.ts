import { types as t } from '@babel/core';

function isComponentName(name: string) {
  return !!name.match(/^[A-Z]/u);
}

export const isComponent = (node: t.FunctionDeclaration) => {
  if (isComponentName(node.id.name)) return true;
  return false;
};
