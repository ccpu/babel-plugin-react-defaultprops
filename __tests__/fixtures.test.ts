import path from 'path';
import glob from 'fast-glob';
import { transform } from './utils';

const manualTestFixturesPath = path.resolve('__tests__/fixtures/manual');

describe('should pass fixtures', () => {
  const files = glob.sync('./fixtures/**/*.{ts,tsx,js,jsx}', {
    cwd: __dirname,
    dot: false,
  });

  files.forEach((basePath) => {
    const filePath = String(basePath);

    if (filePath.includes('/manual/')) {
      return;
    }

    const parsedFile = path.parse(filePath);

    const pathToSnap = path.resolve(
      process.cwd(),
      '__tests__',
      '__snapshots__',
      parsedFile.dir,
      parsedFile.name + parsedFile.ext + '.shot',
    );

    it(`transforms ${parsedFile.dir}/${parsedFile.base}`, () => {
      const result = transform(
        path.join(__dirname, filePath),
        {
          comments: false,
          value: 'foo',
        },
        {
          presets: [
            '@babel/preset-env',
            '@babel/preset-react',
            '@babel/preset-typescript',
          ],
        },
      );
      expect(result).toMatchSpecificSnapshot(pathToSnap);
    });
  });

  it('should pass empty file', () => {
    const result = transform(
      path.join(manualTestFixturesPath, 'empty-file.ts'),
      {},
      {
        presets: [
          '@babel/preset-env',
          '@babel/preset-react',
          '@babel/preset-typescript',
        ],
      },
    );

    expect(result).toBe('"use strict";');
  });
});
