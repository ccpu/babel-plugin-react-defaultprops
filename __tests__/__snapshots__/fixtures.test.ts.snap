// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass fixtures transforms ./fixtures/with-default-parameters.tsx 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.Component = Component;

function Component(_ref) {
  var bar = _ref.bar,
      _ref$foo = _ref.foo,
      foo = _ref$foo === void 0 ? 'hello' : _ref$foo;
  return bar + foo;
}

Component.__defaultProps = {
  foo: 'hello'
};"
`;
